pkg/infradb/infradb.go

```diff
@@ -0,0 +1,1443 @@
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb/subscriberframework/eventbus"
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb/taskmanager"
+	"github.com/opiproject/opi-evpn-bridge/pkg/storage"
+	"github.com/philippgille/gokv"
+)
+
+var infradb *InfraDB
+var globalLock sync.Mutex
+
+// InfraDB structure
+type InfraDB struct {
+	client gokv.Store
+}
+
+var (
+	// ErrKeyNotFound error for missing key
+	ErrKeyNotFound = errors.New("key not found")
+	// ErrComponentNotFound error for missing component
+	ErrComponentNotFound = errors.New("component not found")
+	// ErrVrfNotFound error for missing vrf
+	ErrVrfNotFound = errors.New("the referenced VRF has not been found")
+	// ErrLogicalBridgeNotFound error for missing logical bridge
+	ErrLogicalBridgeNotFound = errors.New("the referenced Logical Bridge has not been found")
+	// ErrVrfNotEmpty vrf is not empty
+	ErrVrfNotEmpty = errors.New("the VRF is not empty")
+	// ErrLogicalBridgeNotEmpty logical bridge is not empty
+	ErrLogicalBridgeNotEmpty = errors.New("the LogicalBridge is not empty")
+	// ErrRoutingTableInUse routing table is in use
+	ErrRoutingTableInUse = errors.New("the routing table is already in use")
+	// ErrVniInUse vni is in use
+	ErrVniInUse = errors.New("the VNI is already in use")
+	// Add more error constants as needed
+)
+
+// NewInfraDB initializes an InfraDB object
+func NewInfraDB(address string, dbtype string) error {
+	store, err := storage.NewStore(dbtype, address)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	infradb = &InfraDB{
+		client: store.GetClient(),
+	}
+	return nil
+}
+
+// Close closes a infradb connection to the DB
+func Close() error {
+	return infradb.client.Close()
+}
+
+// CreateLB creates an infradb logical bridge object
+func CreateLB(lb *LogicalBridge) error {
+	globalLock.Lock()
+	defer globalLock.Unlock()
+
+	vpns := make(map[uint32]bool)
+
+	subscribers := eventbus.EBus.GetSubscribers("logical-bridge")
+	if subscribers == nil {
+		log.Println("CreateLB(): No subscribers for Logical Bridge objects")
+	}
+
+	log.Printf("CreateLB(): Create Logical Bridge: %+v\n", lb)
+
+	// Check if VNI is already used
+	if lb.Spec.Vni != nil {
+		found, err := infradb.client.Get("vpns", &vpns)
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+
+		if !found {
+			vpns[*lb.Spec.Vni] = false
+		} else {
+			_, ok := vpns[*lb.Spec.Vni]
+			if ok {
+				log.Printf("CreateLB(): VNI already in use: %+v\n", lb.Spec.Vni)
+				return ErrVniInUse
+			}
+			vpns[*lb.Spec.Vni] = false
+		}
+	}
+
+	err := infradb.client.Set(lb.Name, lb)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	// Store VNI to DB in the vpns map
+	if lb.Spec.Vni != nil {
+		err = infradb.client.Set("vpns", &vpns)
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+	}
+
+	// Add the New Created Logical Bridge to the "lbs" map
+	lbs := make(map[string]bool)
+	_, err = infradb.client.Get("lbs", &lbs)
+	if err != nil {
+		log.Println(err)
```

Comment:
do we need to remove fall back and remove set `vpns`
```Go
	err = infradb.client.Set("vpns", &vpns)
		if err != nil {
			log.Println(err)
			return err
		}
```
`lb` and other objects?

Reply:
No we do not believe we should fall back in the operations with the DB. If there is a need for this we can discuss it in the future. We can live with this right now.  I believe if we see any need we can revisit in the future. I would say let's not be too picky with the code right now. Let's make sure that the basic functionality is working currently and we can make more picky changes and enhancements in additional PRs. 

WDYT ?

Reply:
Then at least an issue should be created after merging this PR

Reply:
Yes we can create an issue just to register it as point for future discussions if that is needed just so we do not forget.